### DISCUSSION TITLE
구문(Statement) 와 표현식(expression) 의 차이?

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#25] [CHAPTER 3 람다 표현식 - Q&A] [answered]
(91p)
![image](https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/77484084/fda77ae2-d119-4846-a6a8-c7894aca5d01)

- 자바스크립트에서 값,문,식 이라고 하던데 그 문 , 식 인가?
- 아래 사진처럼 블록스타일인데도 return 없이 sysout만 있는데도 괜찮은 것은 sysout이 구문이기 때문인가? 블록스타일에서도 구문이 있다면 return없어도 되는가?

![image](https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/77484084/c551f97f-cb15-4d4a-a192-5e23ed32d781)

**Comment by corock** [comment, marked as answer]
각각의 정의를 간단하게 알고 있으면 도움이 될 것 같아 저도 이번 기회에 정리해 봤습니다.

- 식(_expression_): 하나 이상의 값(_value_)을 반드시 산출(_produce_)할 수 있는 코드
- 문(_statement_): 무언가를 수행(_action_)할 수 있는 코드 블록; 실행할 수 있는(_executable_) 최소 표현식

식은 표현식이라고도 하며, 리터럴이나 연산자의 조합으로 만들어집니다.
그래서 _statement_ 는 _expression_ 의 부분 집합으로 좀 더 큰 범주에 속합니다.

`System.out.println(String)` 은 입력받은 문자열을 출력할 수 있는 _statement_ 이며,
`return` 예약어가 없다면 `void` 타입으로 `Consumer<T>` 등의 함수형 인터페이스를 사용할 수 있습니다.

**Reply by Irisation23** [reply]
`statement`란 의미는 코드 블록 사용을 의미하는 것으로 알고 있습니다.
즉 `{}` 코드 블록으로 코드를 작성한다는 의미는 한줄 한줄이 컴파일이 가능해야 한다는 것을 의미함으로

아래의 코드를 확인 해 보면 비교 가능하다 생각합니다.
- `{}`(statement) 제대로 지키지 못한 경우
```java
public String makeIronMan(String literal) {
  "Iron Man" // 이 줄은 뭐죠?
}
```
- 올바르게 지킨 경우
```java
public String makeIronMan(String literal) {
  return "Iron Man"; 
}
```

**Comment by JoisFe** [comment]
C, 자바 등 프로그래밍 언어를 처음 접할때 `statement`, `expression` 등 이러한 keyword 들을 기본서를 통해 접할때 배우게 되지만 
정확학 개념을 까먹게 되는 것 같습니다.

코틀린 공부를 하다가 표현식 이라는 개념이 나왔었는데 그때 느낀 감정이 순간 뭐지? 였습니다.
[참고사항](https://github.com/orgs/Programming-Language-Practice/discussions/10#discussion-4822555)
해당 링크에 예전에 정리를 해두었는데 해당 질문과 적합한 상황의 내용이 아닌가 싶습니다.

**Comment by Youngju-Jang** [comment]
- 값 < 식(expression) < 문(statement)
- 문 : 실행할 수 있는(executable) 최소 표현식
- 식은 표현식이라고도 하며, 리터럴이나 연산자의 조합
    - 리터럴 (Literal): 53, 62, 126 과 같은 숫자 값 자체
    - 연산자 (Operator): +, - 처럼 계산을 수행하는 기호

- 문 은 되어야 혼자 존재가능하다.
